### PR TITLE
add no-accessor-field-mismatch

### LINKED ISSUE #237: Rule : Getters and setters should access the expected fields
Defined in [RSPEC-4275](https://jira.sonarsource.com/browse/RSPEC-4275)

### PR DESCRIPTION
**Pull request by inverno**
Fixes #237

### CODE REVIEW

File: sonarts-core/src/rules/noAccessorFieldMismatchRule.ts [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    method: ts.MethodDeclaration | ts.AccessorDeclaration,
+  ): { type: string; name: string } | undefined {
+    const name = method.name.getText().toLowerCase();
+    if (method.kind === ts.SyntaxKind.SetAccessor) {
+      return { type: "setter", name };
+    }
+    if (method.kind === ts.SyntaxKind.GetAccessor) {
+      return { type: "getter", name };
+    }
+    if (name.startsWith("set") || name.startsWith("Set")) {
```
Comment (vilchik-elena) [inline sonarts-core/src/rules/noAccessorFieldMismatchRule.ts]: @inverno dead code

File: sonarts-core/src/rules/noAccessorFieldMismatchRule.ts [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+  }
+
+  private static matchingField(
+    method: ts.MethodDeclaration | ts.AccessorDeclaration,
+    targetName: string,
+  ): ts.PropertyDeclaration | ts.ParameterDeclaration | undefined {
+    const containingClass = nav.firstAncestor(method, [
+      ts.SyntaxKind.ClassDeclaration,
+      ts.SyntaxKind.ClassExpression,
+    ]) as ts.ClassDeclaration | ts.ClassExpression;
```
Comment (vilchik-elena) [inline sonarts-core/src/rules/noAccessorFieldMismatchRule.ts]: @inverno We can find methods inside object literal as well

File: sonarts-core/src/rules/noAccessorFieldMismatchRule.ts [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    if (
+      accessorIsPublic &&
+      accessor.body &&
+      matchingField &&
+      Walker.bodyIsSimple(accessor.body, setterOrGetter.type) &&
+      !this.fieldIsUsed(accessor, matchingField)
+    ) {
+      this.addFailureAtNode(
+        accessor.name,
+        `Refactor this ${setterOrGetter.type} so that it actually refers to the property '${matchingField.name!.getText()}'`,
```
Comment (SonarTech) [inline sonarts-core/src/rules/noAccessorFieldMismatchRule.ts]: MINOR Forbidden non null assertion [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS2966)

File: sonarts-core/src/rules/noAccessorFieldMismatchRule.ts [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+    } else {
+      return [];
+    }
+  }
+
+  private static bodyIsSimple(body: ts.Block, type: string): boolean {
+    if (body.statements.length > 1) {
+      return false;
+    }
+    if (body.statements.length == 0) {
```
Comment (SonarTech) [inline sonarts-core/src/rules/noAccessorFieldMismatchRule.ts]: MAJOR == should be === [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS1440)

File: sonarts-core/src/rules/noAccessorFieldMismatchRule.ts [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+          return !!expression.getText().startsWith("this.");
+        }
+      }
+      return false;
+    }
+  }
+
+  private fieldIsUsed(method: ts.MethodDeclaration | ts.AccessorDeclaration, field: Field): boolean {
+    const body = method.body;
+    const usage = this.symbols.getUsage(field.name!);
```
Comment (SonarTech) [inline sonarts-core/src/rules/noAccessorFieldMismatchRule.ts]: MINOR Forbidden non null assertion [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS2966)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**